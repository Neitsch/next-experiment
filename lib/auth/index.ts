import Cookie from "js-cookie";
import jwtDecode from "jwt-decode";

const getQueryParams = () => {
  const params = {};
  window.location.href.replace(
    /([^(?|#)=&]+)(=([^&]*))?/g,
    ($0, $1, $2, $3) => {
      params[$1] = $3;
    },
  );
  return params;
};

export const extractInfoFromHash = () => {
  if (!process.browser) {
    return undefined;
  }
  const { id_token, access_token, state, expires_in } = getQueryParams();
  return { token: id_token, access_token, secret: state, expires_in };
};

export const setToken = (token, accessToken, expiresIn) => {
  if (!process.browser) {
    return;
  }
  const expires = new Date();
  expires.setSeconds(expires.getSeconds() + expiresIn);
  Cookie.set("user", jwtDecode(token), { expires, SameSite: "Lax" });
  Cookie.set("jwt", token, { expires, SameSite: "Lax" });
  Cookie.set("token", accessToken, { expires, SameSite: "Lax" });
};

export const unsetToken = () => {
  if (!process.browser) {
    return;
  }
  Cookie.remove("jwt");
  Cookie.remove("user");
  Cookie.remove("secret");

  // to support logging out from all windows
  window.localStorage.setItem("logout", Date.now());
};

export const getUserFromServerCookie = req => {
  const cookie = getServerCookie(req);
  if (!cookie) {
    return undefined;
  }
  return jwtDecode(cookie);
};

export const getServerCookie = req => {
  if (!req.headers.cookie) {
    return undefined;
  }
  const jwtCookie = req.headers.cookie
    .split(";")
    .find(c => c.trim().startsWith("jwt="));
  if (!jwtCookie) {
    return undefined;
  }
  return jwtCookie.split("=")[1];
};

export const getLocalCookie = () => {
  return Cookie.getJSON("jwt");
};

export const getUserFromLocalCookie = () => {
  return Cookie.getJSON("user");
};

export const setSecret = secret =>
  Cookie.set("secret", secret, { SameSite: "Lax" });

export const checkSecret = secret => {
  return Cookie.get("secret") === secret;
};
