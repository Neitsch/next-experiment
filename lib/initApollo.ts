import { Hermes } from "apollo-cache-hermes";
import { persistCache } from "apollo-cache-persist";
import { ApolloClient } from "apollo-client";
import { setContext } from "apollo-link-context";
import { HttpLink } from "apollo-link-http";
import { createPersistedQueryLink } from "apollo-link-persisted-queries";
import isomorphicUnfetch from "isomorphic-unfetch";
import localForage from "localforage";

let apolloClient = null;

export const setAuthorizationLink = jwt =>
  setContext(() => ({
    headers: {
      authorization: "Bearer " + jwt,
    },
  }));

export const setupLink = jwt =>
  setAuthorizationLink(jwt).concat(
    createPersistedQueryLink().concat(
      new HttpLink({
        credentials: "same-origin",
        fetch: isomorphicUnfetch,
        uri: "http://localhost:3000/graphql",
      }),
    ),
  );

function create(jwt) {
  const cache = new Hermes({});
  if (process.browser) {
    persistCache({
      cache,
      storage: localForage,
    });
  }
  return new ApolloClient({
    cache,
    connectToDevTools: !!process.browser,
    link: setAuthorizationLink(jwt).concat(
      createPersistedQueryLink().concat(
        new HttpLink({
          credentials: "same-origin",
          fetch: isomorphicUnfetch,
          uri: "http://localhost:3000/graphql",
        }),
      ),
    ),
    ssrMode: !process.browser,
  });
}

export const resetClient = () => {
  apolloClient = null;
};

export default function initApollo(jwt) {
  if (!process.browser) {
    return create(jwt);
  }
  if (!apolloClient) {
    apolloClient = create(jwt);
  }
  return apolloClient;
}
