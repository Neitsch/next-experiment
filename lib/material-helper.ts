/* tslint:disable-next-line */
import green from "@material-ui/core/colors/green";
/* tslint:disable-next-line */
import purple from "@material-ui/core/colors/purple";
import {
  createGenerateClassName,
  createMuiTheme,
  /* tslint:disable-next-line */
} from "@material-ui/core/styles";
import { SheetsRegistry } from "jss";

const theme = createMuiTheme({
  palette: {
    primary: {
      dark: purple[700],
      light: purple[300],
      main: purple[500],
    },
    secondary: {
      dark: green[700],
      light: green[300],
      main: green[500],
    },
  },
});

function createPageContext() {
  return {
    generateClassName: createGenerateClassName(),
    sheetsManager: new Map(),
    sheetsRegistry: new SheetsRegistry(),
    theme,
  };
}

export function getPageContext() {
  if (!process.browser) {
    return createPageContext();
  }

  // Reuse context on the client-side.
  if (!global.__INIT_MATERIAL_UI__) {
    global.__INIT_MATERIAL_UI__ = createPageContext();
  }

  return global.__INIT_MATERIAL_UI__;
}
