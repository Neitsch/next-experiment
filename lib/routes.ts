import Routes from "next-routes";

const routes = new Routes().add("index", "/");
export default routes;
export const Link = routes.Link;
export const Router = routes.Router;
