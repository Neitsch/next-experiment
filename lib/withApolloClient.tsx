import { ApolloClient } from "apollo-client";
/* tslint:disable-next-line */
import Head from "next/head";
import React from "react";
import { getDataFromTree } from "react-apollo";

import initApollo from "./initApollo";

export default App => {
  return class Apollo extends React.Component {
    public static displayName = "withApollo(App)";
    public static async getInitialProps(ctx) {
      const { Component, router } = ctx;
      let appProps = {};
      if (App.getInitialProps) {
        appProps = await App.getInitialProps(ctx);
      }
      const apolloState = {
        data: {},
      };
      const apollo = initApollo({});
      try {
        await getDataFromTree(
          <App
            {...appProps}
            Component={Component}
            router={router}
            apolloState={apolloState}
            apolloClient={apollo}
          />,
        );
      } catch (error) {
        /* tslint:disable-next-line */
        console.error("Error while running `getDataFromTree`", error);
      }

      if (!process.browser) {
        Head.rewind();
      }
      apolloState.data = apollo.cache.extract();

      return {
        ...appProps,
        apolloState,
      };
    }

    private apolloClient: ApolloClient<any>;

    constructor(props) {
      super(props);
      this.apolloClient =
        props.apolloClient || initApollo(props.apolloState.data);
    }

    public render() {
      return <App {...this.props} apolloClient={this.apolloClient} />;
    }
  };
};
