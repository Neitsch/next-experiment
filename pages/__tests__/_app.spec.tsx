import { shallow } from "enzyme";
import React from "react";

import { getPageContext } from "../../lib/material-helper";
import withApolloClient from "../../lib/withApolloClient";
import App from "../_app";

jest.unmock("../_app");
jest.mock("../../lib/withApolloClient", () => () => a => a);
jest.mock("apollo-client", () => ({
  ApolloClient: ({ children }) => children,
}));
jest.mock("react-apollo", () => ({
  ApolloProvider: ({ children }) => children,
}));
getPageContext.mockReturnValue({
  generateClassName: jest.fn(),
  sheetsManager: new Map(),
  sheetsRegistry: {
    add: jest.fn(),
  },
  theme: {},
});

describe("App page", () => {
  it("renders", () => {
    const app = <App Component={() => <div />} apollo={jest.fn()} />;
    const rendered = shallow(app);
    expect(rendered).toMatchSnapshot();
  });
  it("renders - with JSS", () => {
    const removeChild = jest.fn();
    document.querySelector = jest.fn().mockReturnValue({
      parentNode: {
        removeChild,
      },
    });
    const app = <App Component={() => <div />} apollo={jest.fn()} />;
    const rendered = shallow(app);
    expect(rendered).toMatchSnapshot();
    expect(removeChild).toHaveBeenCalled();
  });
});
