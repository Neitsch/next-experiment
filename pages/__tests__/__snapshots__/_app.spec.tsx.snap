// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App page renders - with JSS 1`] = `
<mockConstructor>
  <JssProvider
    generateClassName={[MockFunction]}
    registry={
      Object {
        "add": [MockFunction],
      }
    }
  >
    <MuiThemeProvider
      sheetsManager={Map {}}
      theme={Object {}}
    >
      <WithStyles(CssBaseline) />
      <ApolloProvider
        client={[MockFunction]}
      >
        <WithStyles(CondorAppBar) />
        <Component
          pageContext={
            Object {
              "generateClassName": [MockFunction],
              "sheetsManager": Map {},
              "sheetsRegistry": Object {
                "add": [MockFunction],
              },
              "theme": Object {},
            }
          }
        />
      </ApolloProvider>
    </MuiThemeProvider>
  </JssProvider>
</mockConstructor>
`;

exports[`App page renders 1`] = `
<mockConstructor>
  <JssProvider
    generateClassName={[MockFunction]}
    registry={
      Object {
        "add": [MockFunction],
      }
    }
  >
    <MuiThemeProvider
      sheetsManager={Map {}}
      theme={Object {}}
    >
      <WithStyles(CssBaseline) />
      <ApolloProvider
        client={[MockFunction]}
      >
        <WithStyles(CondorAppBar) />
        <Component
          pageContext={
            Object {
              "generateClassName": [MockFunction],
              "sheetsManager": Map {},
              "sheetsRegistry": Object {
                "add": [MockFunction],
              },
              "theme": Object {},
            }
          }
        />
      </ApolloProvider>
    </MuiThemeProvider>
  </JssProvider>
</mockConstructor>
`;
